From 1873778253421364e0d8d7d3d89b36b93be578d7 Mon Sep 17 00:00:00 2001
From: WangJianhui <wjh@rock-chips.com>
Date: Thu, 14 Sep 2017 09:01:24 +0800
Subject: [PATCH] glmark2: support for rockchip

Signed-off-by: WangJianhui <wjh@rock-chips.com>
---
 src/native-state-drm.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/native-state-drm.cpp b/src/native-state-drm.cpp
index 4df75ec..d886e34 100644
--- a/src/native-state-drm.cpp
+++ b/src/native-state-drm.cpp
@@ -200,6 +200,7 @@ NativeStateDRM::init()
     // TODO: Replace this with something that explicitly probes for the loaded
     // driver (udev?).
     static const char* drm_modules[] = {
+	"rockchip",
         "i915",
         "nouveau",
         "radeon",
-- 
1.9.1

