From ded650836c6f164a26dedd56db90d8c868ee16ae Mon Sep 17 00:00:00 2001
From: Hans Yang <yhx@rock-chips.com>
Date: Fri, 14 Dec 2018 16:19:52 +0800
Subject: [PATCH 1/1] fix force fullscreen not work

Signed-off-by: Hans Yang <yhx@rock-chips.com>
---
 fftools/ffplay.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/fftools/ffplay.c b/fftools/ffplay.c
index be37fe2..95a2109 100644
--- a/fftools/ffplay.c
+++ b/fftools/ffplay.c
@@ -4357,6 +4357,15 @@ int main(int argc, char **argv)
                 if (!SDL_GetRendererInfo(renderer, &renderer_info))
                     av_log(NULL, AV_LOG_VERBOSE, "Initialized %s renderer.\n", renderer_info.name);
             }
+            if (is_full_screen) {
+                int displayIndex = -1;
+                SDL_Rect rect = { 0, 0, 0, 0};
+                displayIndex = SDL_GetWindowDisplayIndex(window);
+                SDL_GetDisplayBounds(displayIndex, &rect);
+                screen_width = rect.w;
+                screen_height = rect.h;
+                printf("screen_width, screen_height: %d, %d\n", screen_width, screen_height);
+            }
         }
         if (!window || !renderer || !renderer_info.num_texture_formats) {
             av_log(NULL, AV_LOG_FATAL, "Failed to create window or renderer: %s", SDL_GetError());
-- 
2.7.4

